common/src/main/java/com/graphicsfuzz/common/util/PipelineInfo.java

```diff
@@ -251,6 +251,37 @@ public void removeUniformBinding(String uniformName) {
     }
   }
 
+  public int appendValueToUniform(String uniformName, Number value) {
+
+    if (!dictionary.has(uniformName)) {
+      throw new IllegalArgumentException("Uniform declaration not found.");
+    }
+
+    dictionary.getAsJsonObject(uniformName).get("args").getAsJsonArray().add(value);
+
+    return dictionary.getAsJsonObject(uniformName).get("args").getAsJsonArray().size() - 1;
+  }
+
+  public int getUnusedBindingNumber() {
+    int number = 0;
+
+    List<Integer> bindings =
+        dictionary.entrySet().stream()
+            .filter(item -> item.getValue().getAsJsonObject().has("binding"))
+            .map(item -> item.getValue().getAsJsonObject()
+            .get("binding").getAsInt()).collect(Collectors.toList());
+
+    for (Integer binding: bindings) {
+      // If the result of subtraction is bigger than 1, an unused binding number was found.
+      if (binding - number > 1) {
+        break;
+      }
+      number++;
+    }
```

Comment:
I am a bit worried about this method. 

E.g. bindings = [ 1 ]
result is: 1. This binding is used.

bindings = [ 1, 2 ]
result is: 2. This binding is used.

bindings = [ 0 ]
result is: 1. OK.

bindings = [ 0, 1 ]
result is: 2. OK

bindings = [ 2, 0 ]
result is: 0. This binding is used.

I think it would be easier and clearer to iterate from 0 to max int until you find a number that is not used (by checking a set of used binding).

Reply:
That's right. It fails if the numbers are in arbitrary order. Fixed now as suggested.